compose/ui/ui-text/src/webMain/kotlin/androidx/compose/ui/text/Cache.web.kt

```diff
@@ -16,9 +16,83 @@
 
 package androidx.compose.ui.text
 
-// TODO Use WeakMap once available https://youtrack.jetbrains.com/issue/KT-44309
-internal actual typealias WeakKeysCache<K, V> = NoCache<K, V>
+import kotlin.js.ExperimentalWasmJsInterop
+import kotlin.js.JsAny
+import kotlin.js.JsReference
+import kotlin.js.get
+import kotlin.js.toJsReference
+import kotlin.js.unsafeCast
 
-internal class NoCache<K : Any, V> : Cache<K, V> {
-    override fun get(key: K, loader: (K) -> V): V = loader(key)
+// We can't use Js WeakMap https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap
+// because it uses identity equality for keys, but the usage of WeakKeysCache in SkikoParagraph requires that the
+// keys would have structural equality.
+@OptIn(ExperimentalWasmJsInterop::class)
+internal actual class WeakKeysCache<K : Any, V: Any>  {
+
+    private val cache = HashMap<Key<K>, V>()
+
+    // When K is finalized, we'll eventually receive a callback.
+    // And we use it to remove the Key(WeakReference<K>) from the HashMap.
+    private val registry = FinalizationRegistry { keyJsReference ->
+        @Suppress("RemoveExplicitTypeArguments") // frontend glitch for js compilation?
+        val key = keyJsReference.unsafeCast<JsReference<Key<K>>>().get<Key<K>>()
+        cache.remove(key)
+    }
+
+    actual inline fun getOrPut(key: K, loader: (K) -> V): V {
+        // Here we don't run a loop over the cache keys to clear the unavailable (after GC) keys.
+        // The loop would be not ignorably cheap, because it would make a js-interop call for deref on
+        // every iteration - a concern for scrolls with different text items (getOrPut is called often).
+        // Here we rely on FinalizationRegistry to clean the unavailable (after GC) keys.
+        val weakKey = Key(key)
+
+        val existing = cache[weakKey]
+        if (existing != null) return existing
+
+        val value = loader(key)
+        cache[weakKey] = value
+
+        registry.register(key.toJsReference(), weakKey.toJsReference())
+
+        return value
+    }
+
+    internal class Key<K : Any>(key: K) {
+        private val ref = WeakReference(key)
+        private val hash: Int = key.hashCode()
+
+        override fun equals(other: Any?): Boolean {
+            if (this === other) return true
+            other as Key<*>
+            val a = ref.get()
+            val b = other.ref.get()
+            if (a == null || b == null) {
+                // If either side is cleared, they should not be considered equal,
+                // unless they're literally the same Key instance, which is already handled above.
```

Comment:
It's not handled above. `this === other` is about Key class instances, not `WeakReference`s